homepage/content/blog/2021-03-05-switchMap-concatMap-mergemap-exhaustmap-explained.md

```diff
@@ -0,0 +1,264 @@
+    // was just sent, does nothing to the stream!
+    .pipe(
+      tap(value => console.log("--> sent out", value)),
+      [OPERATOR_HERE]((value) => this.anyLongRunningOp(value)))
+    // We console.log the output, which is 'first' or 'second' or
+    .subscribe(value => console.log("<-- received", value));
+
+  // After subscribing we emit the two value in the observable, could also be more than that
+  this.sub.next('first');
+  this.sub.next('second');
+
+  console.log("-------");
+}
+
+anyLongRunningOp(value: string) {
+  return new Observable(observer => {
+    setTimeout(() => {
+      observer.next(value);
+      observer.complete();
+    }, 2000);
+  });
+}
+```
+
+The `fireEvents()` method emits in the `Subject` `sub` two times one value each. First the value `first` and immediately after the value `second`. This value then gets processed by the operator we want to look how it behaves when multiple values come in.
+
+## Similarities
+
+All of those operators can be piped to an observable and return a new observable. They keep the stream. They resolve the value but return a new observable. This makes them higher order mapping operators. You can subscribe to the outcome but initially `switchMap`, `mergeMap`, `concatMap` and `exhaustMap` return an observable which then can be processed further.
+
+## SwitchMap
+
+Let us take the `switchMap` operator first. I explained it already in the previous post but let us take it as the first one and look how it behaves when multiple values come in.
+
+```ts
+fireEvents() {
+
+  // Here we react ot everything which is fired in the subject
+  this.sub
+    // Here we can take the operator we want to take a look at which returns the
+    // result from the anyLongRunningOp method which is the value itself
+    // (for the sake of simplicity). The `tap` only prints out what
+    // was just sent, does nothing to the stream!
+    .pipe(
+      tap(value => console.log("--> sent out", value)),
+      switchMap(value => this.obsService.anyLongRunningOp(value))
+    )
+    // We console.log the output, which is 'first' or 'second' or
+    .subscribe(value => console.log("<-- received", value));
+
+  // After subscribing we emit the two value in the observable, could also be more than that
+  this.sub.next("first");
+  this.sub.next("second");
+
+  console.log("-------");
+}
+```
+
+The `switchMap` operator takes the first value in the stream `first` and calls the `anyLongRunningOp` with it. _Right after_ it did this it receives the second emit with the value `second`. Now it forgets about the response of the first request. It is not waiting for it. Like the HTTP request which could be done here is out and comes back any when but the `switchMap` operator does not care about the first one. It calls the `anyLongRunningOp` with the `second` parameter and waits for _that one_'s answer. And so with multiple ones, it is only interested in the response of the _last_ one it fired. Everything before got ignored.

```

Comment:
Like the HTTP request which could be done here is out and comes back any when but the `switchMap` operator does not care about the first one

Der Satz ist schwierig zu verstehen